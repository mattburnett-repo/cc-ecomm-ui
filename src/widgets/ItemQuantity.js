
import { useSelector, useDispatch } from 'react-redux'

import { selectCurrentCart } from '../features/carts/cartsSlice';

import { changeCurrentCartItemQuantity } from '../features/carts/currentCartActions' 

export default function ItemQuantity (props) {
    const itemId = props.itemId
    const theArray = [1,2,3,4,5,6,7,8,9,10]; // FIXME there has to be a cleaner way to do this
    
    const dispatch = useDispatch()

    const currentCartItems = useSelector(selectCurrentCart)
    const theItem = currentCartItems.find(item => item.id === props.itemId)
    
    let theSelectedOption = 0
    if(theItem) {
        theSelectedOption = theItem.quantity
    } 

    return (
        <>
            <label htmlFor="item-quantity-selector">Quantity:</label>
            <select role="presentation" id="item-quantity-selector" name="itemQuantity" aria-label="item-quantity-selector" 
                    defaultValue={theSelectedOption} onChange={(e) => dispatch(changeCurrentCartItemQuantity(itemId, parseInt(e.target.value)))}>
                {theArray.map(i => (
                    <option key={i} value={i}>{i}</option>
                ))}
            </select>
        </>
    )
}
