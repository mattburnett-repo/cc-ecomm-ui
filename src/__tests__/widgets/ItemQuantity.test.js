import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
// import userEvent from '@testing-library/user-event';

import ItemQuantity  from '../../widgets/ItemQuantity';

describe('<ItemQuantity /> feature tests', () => {
    beforeEach(() => render(<ItemQuantity />));

    it('should render an item-quantity-selector element', () => {
        screen.getByRole('presentation', { name: /item-quantity-selector/i})
    })
    it('should select options', () => {
        window.alert = () => {}; // https://stackoverflow.com/questions/55088482/jest-not-implemented-window-alert
        let theVals = screen.getAllByRole('presentation', {name: /item-quantity-selector/i});

        fireEvent.change(theVals[0], { target: { value: 7 } })
        let options = theVals[0];
        expect(options[0].selected).toBeFalsy();
        expect(options[1].selected).toBeFalsy();
        expect(options[2].selected).toBeFalsy();
        expect(options[3].selected).toBeFalsy();
        expect(options[4].selected).toBeFalsy();
        expect(options[5].selected).toBeFalsy();
        expect(options[6].selected).toBeTruthy();
        expect(options[7].selected).toBeFalsy();
        expect(options[8].selected).toBeFalsy();
        expect(options[9].selected).toBeFalsy();
    })

    it('creates a snapshot', () => {
        const tree = renderer.create(<ItemQuantity /> ).toJSON();
        expect(tree).toMatchSnapshot();  
    })
}); // end feature
