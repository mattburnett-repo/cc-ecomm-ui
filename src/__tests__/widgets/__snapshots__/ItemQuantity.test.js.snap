// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemQuantity /> feature tests creates a snapshot 1`] = `
Array [
  <label
    htmlFor="item-quantity-selector"
  >
    Quantity:
  </label>,
  <select
    aria-label="item-quantity-selector"
    defaultValue={0}
    id="item-quantity-selector"
    name="itemQuantity"
    onChange={[Function]}
    role="presentation"
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={2}
    >
      2
    </option>
    <option
      value={3}
    >
      3
    </option>
    <option
      value={4}
    >
      4
    </option>
    <option
      value={5}
    >
      5
    </option>
    <option
      value={6}
    >
      6
    </option>
    <option
      value={7}
    >
      7
    </option>
    <option
      value={8}
    >
      8
    </option>
    <option
      value={9}
    >
      9
    </option>
    <option
      value={10}
    >
      10
    </option>
  </select>,
]
`;
